File: internal/metastore/advisory_lock.go
package metastore

import (
	"context"
	"database/sql"
	"fmt"
)

// AdvisoryLock wraps MySQL GET_LOCK/RELEASE_LOCK on a dedicated connection.
// The lock is connection-scoped: it is automatically released if the connection
// drops, preventing deadlocks from crashed processes.
type AdvisoryLock struct {
	conn *sql.Conn
	name string
	held bool
}

// AcquireAdvisoryLock obtains a named advisory lock with a timeout.
// Uses db.Conn() to get a dedicated connection (required for lock affinity).
func AcquireAdvisoryLock(ctx context.Context, db *sql.DB, name string, timeoutSeconds int) (*AdvisoryLock, error) {
	conn, err := db.Conn(ctx)
	if err != nil {
		return nil, fmt.Errorf("advisory lock conn: %w", err)
	}

	var result sql.NullInt64
	err = conn.QueryRowContext(ctx, "SELECT GET_LOCK(?, ?)", name, timeoutSeconds).Scan(&result)
	if err != nil {
		conn.Close()
		return nil, fmt.Errorf("GET_LOCK(%s): %w", name, err)
	}

	if !result.Valid || result.Int64 != 1 {
		conn.Close()
		return nil, fmt.Errorf("GET_LOCK(%s): timeout or error (result=%v)", name, result)
	}

	return &AdvisoryLock{conn: conn, name: name, held: true}, nil
}

// Release releases the advisory lock and closes the connection.
func (l *AdvisoryLock) Release(ctx context.Context) error {
	if !l.held {
		return nil
	}
	defer l.conn.Close()
	l.held = false

	var result sql.NullInt64
	err := l.conn.QueryRowContext(ctx, "SELECT RELEASE_LOCK(?)", l.name).Scan(&result)
	if err != nil {
		return fmt.Errorf("RELEASE_LOCK(%s): %w", l.name, err)
	}
	return nil
}

File: crates/metalog-metastore/src/advisory_lock.rs
use sqlx::MySqlPool;

/// Connection-scoped MySQL advisory lock via `GET_LOCK` / `RELEASE_LOCK`.
///
/// The lock is tied to a dedicated connection (not the pool). If the connection
/// drops, the lock is automatically released server-side, preventing deadlocks
/// from crashed processes.
pub struct AdvisoryLock {
    conn: sqlx::pool::PoolConnection<sqlx::MySql>,
    name: String,
    held: bool,
}

/// Error returned when the advisory lock cannot be acquired.
#[derive(Debug, thiserror::Error)]
pub enum AdvisoryLockError {
    #[error("advisory lock {0}: not acquired (timeout or held by another session)")]
    NotAcquired(String),

    #[error("advisory lock {0}: {1}")]
    Sql(String, #[source] sqlx::Error),
}

impl AdvisoryLock {
    /// Acquires a named advisory lock with a timeout (in seconds).
    ///
    /// Uses `pool.acquire()` to get a dedicated connection (required for lock
    /// affinity — the lock is released when the connection closes).
    pub async fn acquire(
        pool: &MySqlPool,
        name: &str,
        timeout_seconds: i32,
    ) -> Result<Self, AdvisoryLockError> {
        let mut conn = pool
            .acquire()
            .await
            .map_err(|e| AdvisoryLockError::Sql(name.into(), e))?;

        let result: Option<i64> = sqlx::query_scalar("SELECT GET_LOCK(?, ?)")
            .bind(name)
            .bind(timeout_seconds)
            .fetch_one(&mut *conn)
            .await
            .map_err(|e| AdvisoryLockError::Sql(name.into(), e))?;

        match result {
            Some(1) => Ok(Self {
                conn,
                name: name.to_string(),
                held: true,
            }),
            _ => Err(AdvisoryLockError::NotAcquired(name.into())),
        }
    }

    /// Releases the advisory lock and returns the connection to the pool.
    ///
    /// The lock is always released when the connection closes, so `held` is set
    /// to false after the connection is dropped to ensure consistent state.
    pub async fn release(&mut self) -> Result<(), AdvisoryLockError> {
        if !self.held {
            return Ok(());
        }

        let release_result: Result<Option<i64>, _> = sqlx::query_scalar("SELECT RELEASE_LOCK(?)")
            .bind(&self.name)
            .fetch_one(&mut *self.conn)
            .await;

        // Mark as released regardless of RELEASE_LOCK result — dropping the
        // connection definitively releases the lock server-side.
        self.held = false;

        if let Err(e) = release_result {
            return Err(AdvisoryLockError::Sql(self.name.clone(), e));
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn error_display() {
        let err = AdvisoryLockError::NotAcquired("test_lock".into());
        assert!(err.to_string().contains("not acquired"));
    }
}
